### ISSUE TITLE
np.place warning: WRITEBACKIFCOPY detected in array_dealloc

### ISSUE THREAD

**Issue by ViaFerrata** [numpy/numpy#12806]
Hello there,

I'm having a problem with getting a runtime warning:

/.../.virtualenv/python_3_env/lib/python3.6/site-packages/numpy/lib/function_base.py:1725: RuntimeWarning: WRITEBACKIFCOPY detected in array_dealloc.  Required call to PyArray_ResolveWritebackIfCopy or PyArray_DiscardWritebackIfCopy is missing.

Actually, I'm curious about what this warning actually means?
The origin of the warning is from a numpy.place, wich is placed in a generator:

```
def mygen():
    ...
    yield np.place(arr, bool_mask, vals)
```

If I remove the np.place operation, there is no warning.
The weird thing is, that I only get this warning, after I call the generator for the second time, e.g.

```
a = mygen() # no warning
a = mygen() # warning
a = mygen() # no warnings more from here on
...

```

I'm using Python 3.6.2 and numpy 1.16.0.

**Comment by ViaFerrata** [comment]
Thank you for the quick answer.
I'd have tried to make a MWE, but the generator is hidden deep in a framework for training neural nets with Keras & TF (the generator yields batches of images and labels to the net). So it's not that easy.

Here is a more detailed scheme of how it looks:

```
import h5py
import numpy as np
import keras

def mygen(filename):
    # load input from a chunked hdf5 file
    f = h5py.File(filename, 'r')

    n_entries = 0
    while ... :
        ... # some other things
        output = {} # this will be yielded

        # load input data, a slice of a structured np array
        struc_arr = f['struc_arr'][n_entries:n_entries+batchsize]

        col_1, col_2 = struc_arr['col_1'], struc_arr['col_2']
        bool_mask = np.logical_and(np.abs(col_1) == 0, col_2 == 1)

        col_3_mod = struc_arr[bool_mask]['col_3'] * struc_arr[bool_mask]['col_4']
        # modify content in col_3
        np.place(struc_array['col_3'], bool_mask, col_3_mod)

        output['col_3'] = struc_arr['col_3']

        n_entries += batchsize
        yield output

    f.close()

# model is a Keras nn model, fit_generator trains it for one epoch
# so the fit_generator calls a lot of nexts on the gen
model.fit_generator(mygen(filename), ...) # no warning during this

... # do something else
model.fit_generator(mygen(filename), ...) # One (!) warning here, at the first yield of the gen

... # do something else
# after this point, no futher warnings, no matter how often this is repeated
model.fit_generator(mygen(filename), ...) 

```

Although I don't think that the original code may really help you, here's the generator that I use:
https://github.com/ViaFerrata/OrcaNet/blob/rework_config/orcanet/utilities/nn_utilities.py#L56

With the function that does the np.place:
https://github.com/ViaFerrata/OrcaNet/blob/rework_config/orcanet/utilities/nn_utilities.py#L74
https://github.com/ViaFerrata/OrcaNet/blob/rework_config/orcanet/utilities/nn_utilities.py#L251

**Comment by seberg** [comment]
Just to note, it is here: https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/compiled_base.c#L332 although I think the error paths probably need some similar handling. In fact, all of those should probably just unset the writebackifcopy flag I guess. Not quite sure what the pattern for that was without checking.

**Comment by seberg** [comment]
@ViaFerrata just to note for you. Even on python implementations that do need the improved behaviour (pypy) there is no actual bug happening. You can safely ignore this warning.
